### DISCUSSION TITLE
Programmatically terminate runs

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8509] [Q&A]
is there a way to programmatically terminate runs? I’m trying to figure out how to build a sensor that will terminate runs if they are taking longer than expected

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1635182950100400?thread_ts=1635182950.100400&cid=C01U954MEER).*

**Comment by yuhan** [comment]
You can use GraphQL API for this. Here's an example query for [Terminate an in-progress run](https://docs.dagster.io/concepts/dagit/graphql#terminate-an-in-progress-run).


Or, you can also cancel a run via the following dagster instance methods:
```python
if instance.run_coordinator.can_cancel_run(run_id):
    instance.run_coordinator.cancel_run(run_id)

# fetch the run to see if its state has changed since calling cancel
run = instance.get_run_by_id(run_id)
if not run.is_finished:
    # just mark run as canceled
    instance.report_run_canceled(run)
```
